core/src/main/scala/kafka/tools/MirrorMaker.scala

```diff
@@ -432,15 +432,20 @@ object MirrorMaker extends Logging with KafkaMetricsGroup {
               trace("Sending message with value size %d and offset %d".format(data.value.length, data.offset))
```

Comment:
The scenario described above will not happen. In order for rebalance to complete in step 2, MirrorMakerThread needs to call `mirrorMakerConsumer.notifyCommit() ` at line 445 after rebalance is triggered, it means that message `m` should have been produced and flushed by MirrorMakerThread before its offset can be committed.

I think the patch should have completely solved the synchronization problem. The root cause of the reordering is that two different threads may be executing `maybeFlushAndCommitOffsets ()`. This patch solves the problem by making sure that only the MirrorMakeThread will execute `maybeFlushAndCommitOffsets()`.